### ISSUE TITLE
netns-nft test fails if nft tables exist

### ISSUE THREAD

**Issue by adrianreber** [checkpoint-restore/criu#1416]
On Cirrus CI and Jenkins where tests are running on native CentOS 8 systems the `netns-nft` fails with:
```
======================= Run zdtm/static/netns-nft in uns =======================
Start test
Test is SUID
./netns-nft --pidfile=netns-nft.pid --outfile=netns-nft.out --filename=netns-nft.test
Run criu dump
Run criu lazy-pages
Run criu restore
Send the 15 signal to  104
Wait for zdtm/static/netns-nft(104) to die for 0.100000
############### Test zdtm/static/netns-nft FAIL at result check ################
Test output: ================================
Error: No such file or directory
list table inet netns-nft-zdtm-test
                ^^^^^^^^^^^^^^^^^^^
04:01:35.029:     4: FAIL: netns-nft.c:51: Can't get nft table (errno = 11 (Resource temporarily unavailable))

 <<< ================================
```

The CentOS 8 systems I have seen actually have nft tables configured, maybe that is something which the test does not expect.

For the Cirrus CI CentOS 8 test I have been doing `mv /usr/sbin/nft /usr/sbin/nft.away` to skip the test, but a real fix would be nice.

@mihalicyn any ideas how to fix that?

**Comment by mihalicyn** [comment]
Please, assign me. The problem is that in recent distributions of CentOS there is no nftables statically linked library.
They lives us with no choice - we have to use dynamically linked libraries.

**Comment by mihalicyn** [comment]
I've submitted an issue about CentOS 8 nftables package.

https://bugs.centos.org/view.php?id=18151

On my Fedora 33 CRIU builds fine with dynamic nftables library and `netns-nft` test passes.

**Comment by adrianreber** [comment]
So test would need to check if CRIU is built with nft support or not. If nft support is not included in CRIU the test should be skipped.

Maybe a feature check that returns false if not linked against nftables?

**Comment by mihalicyn** [comment]
@adrianreber yeah.

We can use something like `ldd ../criu/criu | grep nftables` in checkskip file for `netns-nft` test, but this will only work for the dynamically linked way. If the nftables library linked statically then we have no fully reliable way to detect that.
I think it will be okay just to check `pkg-config --libs libnftables`. Because in our `Makefile` we detecting the nftables library in user system using `pkg-config`.

**Comment by adrianreber** [comment]
> @adrianreber yeah.
> 
> We can use something like `ldd ../criu/criu | grep nftables` in checkskip file for `netns-nft` test, but this will only work for the dynamically linked way. If the nftables library linked statically then we have no fully reliable way to detect that.

I do not think CRIU's build system supports static linking. So yes, your idea would probably work.

> I think it will be okay just to check `pkg-config --libs libnftables`. Because in our `Makefile` we detecting the nftables library in user system using `pkg-config`.

Also correct. It assumes, however, that we build CRIU on the same system as we run the tests. That is not true for how I use zdtm for gating in Fedora/RHEL. I would prefer using `--feature` to have an easy way to query CRIU if it supports it or not, but in the end it is up to @avagin what he wants to merge.

**Comment by mihalicyn** [comment]
It's not a problem - if we want, then I'm ready to prepare PR with the feature for nftables.

**Comment by mihalicyn** [comment]
>I do not think CRIU's build system supports static linking. 

Yes, we have no support to build fully-static CRIU binary. But the nftables can be compiled statically or dynamically depends on which library is present on the user system (it depends on flags which `pkg-config --libs libnftables` returns on a particular system).

**Comment by mihalicyn** [comment]
@adrianreber The only thing that confuses me in the idea to add `criu check` feature is that our mechanism with features serves not so much to test the features of the CRIU itself but to test the features of the kernel.

**Comment by adrianreber** [comment]
> @adrianreber The only thing that confuses me in the idea to add `criu check` feature is that our mechanism with features serves not so much to test the features of the CRIU itself but to test the features of the kernel.

We have at least two checks which do not check kernel features: compat and restore into network namespace

We already use `--feature` to check enable runtime checking of CRIU features. It is not completely new.

**Comment by Snorch** [comment]
Please note that --feature's are normally cached with kerndat cache and if someone would install proper libraries kdat cache can still report libs missing.
